src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Handlers/Nl2Fx/Nl2FxLanguageServerOperationHandler.cs

```diff
@@ -0,0 +1,132 @@
+{
+    /// <summary>
+    /// An abstract handler for NL2FX operations.
+    /// This is a class and not interface to allow us to add new methods in future without breaking existing implementations.
+    /// </summary>
+    public class Nl2FxLanguageServerOperationHandler : ILanguageServerOperationHandler
+    {
+        public bool IsRequest => true;
+
+        public string LspMethod => CustomProtocolNames.NL2FX;
+
+        private readonly INLHandlerFactory _nLHandlerFactory;
+
+        private CustomNL2FxParams _nl2FxRequestParams;
+
+        private NLHandler _nlHandler;
+
+        private CustomNL2FxResult _nl2FxResult;
+
+        private NL2FxParameters _nl2FxParameters;
+
+        public Nl2FxLanguageServerOperationHandler(INLHandlerFactory nLHandlerFactory)
+        {
+            _nLHandlerFactory = nLHandlerFactory;
+        }
+
+        /// <summary>
+        /// Performs pre-handle operations for NL2FX.
+        /// </summary>
+        /// <param name="operationContext"> Language Server Operation Context. </param>
+        private void PreHandleNl2Fx(LanguageServerOperationContext operationContext)
+        {
+            _nlHandler = _nLHandlerFactory?.GetNLHandler(operationContext.GetScope(_nl2FxRequestParams.TextDocument.Uri), _nl2FxRequestParams) ?? throw new NullReferenceException("No suitable handler found to handle Nl2Fx");
+            if (!_nlHandler.SupportsNL2Fx)
+            {
+                throw new NotSupportedException("Nl2fx is not supported");
+            }
+
+            var check = operationContext.Check(_nl2FxRequestParams.TextDocument.Uri, LanguageServer.Nl2FxDummyFormula) ?? throw new NullReferenceException("Check result was not found for NL2Fx operation");
+            var summary = check.ApplyGetContextSummary();
+            _nl2FxParameters = new NL2FxParameters
+            {
+                Sentence = _nl2FxRequestParams.Sentence,
+                SymbolSummary = summary,
+                Engine = check.Engine
+            };
+         }
+
+        /// <summary>
+        /// Performs Core NL2FX operation.
+        /// </summary>
+        /// <param name="operationContext">Language Server Operation Context.</param>
+        /// <param name="cancellationToken">Cancellation Token.</param>
+        /// <returns> Nl2Fx Handle Context extended with NL2Fx result. </returns>
+        private async Task Nl2FxAsync(LanguageServerOperationContext operationContext, CancellationToken cancellationToken)
+        {
+            _nl2FxResult = await _nlHandler.NL2FxAsync(_nl2FxParameters, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Performs post-handle operations for NL2FX.
+        /// </summary>
+        /// <param name="operationContext">Language Server Operation Context.</param>
+        private void PostHandleNl2FxResults(LanguageServerOperationContext operationContext)
+        {
+            var nl2FxResult = _nl2FxResult;

```

Comment:
>  var nl2FxResult = _nl2FxResult; [](http://example.com/codeflow?start=10&length=33)

If we want to disable FinalCheck,  we could add a policy flag on _nlHandler (which host provides) and check that.